File: src/connectors/claude-code.ts
import fs from "fs/promises";
import path from "path";
import fg from "fast-glob";
import type { RawConversation, ConnectorConfig } from "../kb/types.js";
import { BaseConnector } from "./base.js";

export class ClaudeCodeConnector extends BaseConnector {
  name = "claude_code";
  private basePath: string;

  constructor(config: ConnectorConfig) {
    super(config);
    this.basePath = (config.path as string) || path.join(process.env.HOME || "~", ".claude");
  }

  async validate(): Promise<boolean> {
    try {
      await fs.access(this.basePath);
      return true;
    } catch {
      return false;
    }
  }

  async fetch(since?: Date): Promise<RawConversation[]> {
    const conversations: RawConversation[] = [];

    // Read conversation JSONL files from ~/.claude/projects/
    const projectsDir = path.join(this.basePath, "projects");
    try {
      await fs.access(projectsDir);
    } catch {
      return conversations;
    }

    const jsonlFiles = await fg("**/*.jsonl", {
      cwd: projectsDir,
      absolute: true,
    });

    for (const file of jsonlFiles) {
      try {
        const stat = await fs.stat(file);
        if (since && stat.mtime < since) continue;

        const content = await fs.readFile(file, "utf-8");
        const lines = content.trim().split("\n").filter(Boolean);

        const messages = [];
        for (const line of lines) {
          try {
            const msg = JSON.parse(line);
            if (msg.type === "human" || msg.type === "assistant") {
              const textContent = typeof msg.message?.content === "string"
                ? msg.message.content
                : Array.isArray(msg.message?.content)
                  ? msg.message.content
                      .filter((b: { type: string }) => b.type === "text")
                      .map((b: { text: string }) => b.text)
                      .join("\n")
                  : "";

              if (textContent) {
                messages.push({
                  role: msg.type === "human" ? "user" as const : "assistant" as const,
                  content: textContent,
                  timestamp: msg.timestamp,
                });
              }
            }
          } catch {
            // skip malformed lines
          }
        }

        if (messages.length > 0) {
          conversations.push({
            source: "claude_code",
            id: path.basename(file, ".jsonl"),
            messages,
            timestamp: stat.mtime.toISOString(),
            metadata: { file_path: file },
          });
        }
      } catch {
        // skip unreadable files
      }
    }

    // Also read memory files
    const memoryFiles = await fg("**/memory/**/*.md", {
      cwd: this.basePath,
      absolute: true,
    });

    for (const file of memoryFiles) {
      try {
        const stat = await fs.stat(file);
        if (since && stat.mtime < since) continue;

        const content = await fs.readFile(file, "utf-8");
        conversations.push({
          source: "claude_code_memory",
          id: `memory-${path.basename(file, ".md")}`,
          messages: [{ role: "system" as const, content }],
          timestamp: stat.mtime.toISOString(),
          metadata: { file_path: file, type: "memory" },
        });
      } catch {
        // skip
      }
    }

    return conversations;
  }
}

File: synesis/connectors/claude_code.py
from __future__ import annotations

import glob
import json
import os
from datetime import datetime
from pathlib import Path

from synesis.connectors.base import BaseConnector
from synesis.kb.types import ConversationMessage, RawConversation


class ClaudeCodeConnector(BaseConnector):
    name = "claude_code"

    def __init__(self, config: dict):
        super().__init__(config)
        self.base_path = Path(config.get("path", os.path.expanduser("~/.claude")))

    def validate(self) -> bool:
        return self.base_path.exists()

    def fetch(self, since: str | None = None) -> list[RawConversation]:
        since_dt = datetime.fromisoformat(since) if since else None
        conversations: list[RawConversation] = []

        # Read conversation JSONL files
        projects_dir = self.base_path / "projects"
        if projects_dir.exists():
            for jsonl_file in glob.glob(str(projects_dir / "**/*.jsonl"), recursive=True):
                path = Path(jsonl_file)
                mtime = datetime.fromtimestamp(path.stat().st_mtime)
                if since_dt and mtime < since_dt:
                    continue

                messages = self._parse_jsonl(path)
                if messages:
                    conversations.append(
                        RawConversation(
                            source="claude_code",
                            id=path.stem,
                            messages=messages,
                            timestamp=mtime.isoformat(),
                            metadata={"file_path": str(path)},
                        )
                    )

        # Read memory files
        for md_file in glob.glob(str(self.base_path / "**/memory/**/*.md"), recursive=True):
            path = Path(md_file)
            mtime = datetime.fromtimestamp(path.stat().st_mtime)
            if since_dt and mtime < since_dt:
                continue

            content = path.read_text(encoding="utf-8")
            conversations.append(
                RawConversation(
                    source="claude_code_memory",
                    id=f"memory-{path.stem}",
                    messages=[ConversationMessage(role="system", content=content)],
                    timestamp=mtime.isoformat(),
                    metadata={"file_path": str(path), "type": "memory"},
                )
            )

        return conversations

    def _parse_jsonl(self, path: Path) -> list[ConversationMessage]:
        messages = []
        for line in path.read_text(encoding="utf-8").strip().split("\n"):
            if not line.strip():
                continue
            try:
                msg = json.loads(line)
                if msg.get("type") not in ("user", "assistant"):
                    continue

                content = msg.get("message", {}).get("content", "")
                if isinstance(content, list):
                    content = "\n".join(
                        b.get("text", "") for b in content if b.get("type") == "text"
                    )

                if content:
                    messages.append(
                        ConversationMessage(
                            role=msg["type"],  # "user" or "assistant"
                            content=content,
                            timestamp=msg.get("timestamp"),
                        )
                    )
            except (json.JSONDecodeError, KeyError):
                continue
        return messages
